lib/discourse_activity_pub/bulk/process.rb

```diff
@@ -0,0 +1,761 @@
+            ap_type: actor.ap_type,
+            model_type: "User",
+            model_id: user.id,
+          }
+        end
+
+        if update_actors.any?
+          DiscourseActivityPubActor.upsert_all(
+            update_actors,
+            unique_by: %i[ap_id],
+            update_only: %i[model_type model_id],
+          )
+        end
+      end
+
+      def create_posts_from_objects(first_post_object_attrs, reply_object_attrs)
+        result.posts_by_object_ap_id = {}
+        result.topics_by_collection_ap_id = {}
+
+        objects_by_ap_id = create_posts(first_post_object_attrs, "first_post")
+
+        log("created", "topics", created_topics_count) if created_topics_count > 0
+
+        if actor.model.activity_pub_full_topic
+          reply_object_attrs = update_reply_object_attrs_from_stored(reply_object_attrs)
+          reply_objects_by_ap_id = create_posts(reply_object_attrs, "reply")
+          objects_by_ap_id = objects_by_ap_id.merge(reply_objects_by_ap_id)
+        end
+
+        result.objects_by_ap_id = objects_by_ap_id
+
+        log("created", "replies", created_replies_count) if created_replies_count > 0
+
+        update_post_associations_and_remove_orphaned_objects
+      end
+
+      def create_posts(object_attrs, type)
+        create_post_opts_by_ap_id = {}
+
+        object_attrs.each do |attrs|
+          if attrs[:create_post_opts]
+            create_post_opts_by_ap_id[attrs[:ap_id]] = attrs.delete(:create_post_opts)
+          end
+        end
+
+        objects_by_ap_id = store_objects(object_attrs)
+
+        result.activities.each do |activity|
+          object_ap_id = activity.object.json[:id]
+
+          if objects_by_ap_id[object_ap_id]
+            activity.object.stored = objects_by_ap_id[object_ap_id]
+            post = create_post(activity, create_post_opts_by_ap_id)
+            objects_by_ap_id.delete(object_ap_id) unless post
+          end
+        end
+
+        objects_by_ap_id
+      end
+
+      def create_post(activity, create_post_opts_by_ap_id)
+        object = activity.object
+        post = object.stored.model
+
+        if !post
+          create_post_opts = create_post_opts_by_ap_id[object.stored.ap_id] || {}
+          post_actor_ap_id =
+            object.attributed_to ? object.attributed_to.stored.ap_id : actor.stored.ap_id
+
+          post =
+            DiscourseActivityPub::PostHandler.create(
+              result.users_by_actor_ap_id[post_actor_ap_id],
+              object.stored,
+              **create_post_opts.merge(category_id: actor.model.id, import_mode: true),
+            )
+
+          if post.present?
+            if create_post_opts[:topic_id]
+              @created_replies_count += 1
+            else
+              @created_topics_count += 1
+            end
+          end
+        end
+
+        if post.present?
+          result.posts_by_object_ap_id[object.stored.ap_id] = post
+
+          if actor.model.activity_pub_full_topic
```

Comment:
Does this mean that source and target models need to match? It looks like it from my testing so far, if I switch the setting on a target category to "Full topic", running the bulk process raises an error on line 48 just below. 

(More of a general comment for a follow up.)

Reply:
I'm going to investigate this and get back to you.

Reply:
@pmusaraj I've added some fixes and tests for this scenario https://github.com/discourse/discourse-activity-pub/pull/72/commits/1173eaa35f671b4140f73f0bcabf2bc8d8fbc0bc